src/stream.rs

```diff
@@ -0,0 +1,263 @@
+            // whose contract is to make it well-formed UTF-8.
+            let buffer_slice = unsafe { str::from_utf8_unchecked(buffer_slice) };
+            let (bytes_copied, _) = copy_maximum_utf8_prefix(buffer_slice, dst);
+            self.unused_buffer_slice.start += bytes_copied;
+            Ok(bytes_copied)
+        } else if !self.reached_stream_eof {
+            let bytes_in_buffer = ReadBytes::read(&mut self.stream, buffer)?;
+            if bytes_in_buffer == 0 {
+                self.reached_stream_eof = true;
+                let (_, _, written, _) = self.decoder.decode_to_utf8(b"", dst, true);
+                // FIXME: deal with CoderResult::OutputFull here
+                Ok(written)
+            } else {
+                let (_, bytes_read, bytes_written, _) = self.decoder.decode_to_utf8(
+                    &buffer[..bytes_in_buffer], dst, false);
+                self.unused_buffer_slice = bytes_read..bytes_in_buffer;
+                Ok(bytes_written)
+            }
+        } else {
+            Ok(0)
+        }
+    }
+}
+
+pub struct ReadEncoder<Stream: ReadUnicode, Buffer: AsMut<str>> {
+    encoder: Encoder,
+    stream: Stream,
+    reached_stream_eof: bool,
+    buffer: Buffer,
+    unused_buffer_slice: Range<usize>,
+}
+
+impl<Stream: ReadUnicode, Buffer: AsMut<str>> ReadBytes for ReadEncoder<Stream, Buffer> {
+    fn read(&mut self, dst: &mut [u8]) -> io::Result<usize> {
+        let buffer = self.buffer.as_mut();
+        if self.unused_buffer_slice.end > self.unused_buffer_slice.start {
+            let buffer_slice = &buffer[self.unused_buffer_slice.clone()];
+            let bytes_copied = ReadBytes::read(&mut buffer_slice.as_bytes(), dst)?;
+            self.unused_buffer_slice.start += bytes_copied;
+            Ok(bytes_copied)
+        } else if !self.reached_stream_eof {
+            let bytes_in_buffer = ReadUnicode::read(&mut self.stream, buffer)?;
+            if bytes_in_buffer == 0 {
+                self.reached_stream_eof = true;
+                let (_, _, written, _) = self.encoder.encode_from_utf8("", dst, true);
+                // FIXME: deal with CoderResult::OutputFull here
+                Ok(written)
+            } else {
+                let (_, bytes_read, bytes_written, _) = self.encoder.encode_from_utf8(
+                    &buffer[..bytes_in_buffer], dst, false);
+                self.unused_buffer_slice = bytes_read..bytes_in_buffer;
+                Ok(bytes_written)
+            }
+        } else {
+            Ok(0)
+        }
+    }
+}
+
+pub struct WriteDecoder<Stream: WriteUnicode, Buffer: AsMut<[u8]>> {
+    decoder: Decoder,
+    stream: Stream,
+    buffer: Buffer,
+}
+
+impl<Stream: WriteUnicode, Buffer: AsMut<[u8]>> WriteBytes for WriteDecoder<Stream, Buffer> {
+    fn write(&mut self, mut src: &[u8]) -> io::Result<usize> {
+        let buffer = self.buffer.as_mut();
+        let mut written_total = 0;
+        while !src.is_empty() {
```

Comment:
Instead of stopping when the source slice is empty, conversion looks should stop when `Decoder`/`Encoder` returns `InputEmpty`. Otherwise, the caller may fail to flush out residual state in the converter. (In the `Encoder` case, this only applies to ISO-2022-JP. Yay ISO-2022-JP.)